core/trino-main/src/main/java/io/trino/cache/CanonicalSubplan.java

```diff
@@ -0,0 +1,431 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.cache;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.ImmutableBiMap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import io.trino.metadata.TableHandle;
+import io.trino.spi.cache.CacheColumnId;
+import io.trino.spi.cache.CacheTableId;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.connector.SortOrder;
+import io.trino.spi.predicate.TupleDomain;
+import io.trino.sql.ir.Expression;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.optimizations.SymbolMapper;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.TopNRankingNode.RankingType;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterables.getLast;
+import static io.trino.cache.CanonicalSubplanExtractor.columnIdToSymbol;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class represents a canonical subplan. Canonical subplan nodes can either be
+ * leaf nodes ({@link CanonicalSubplan#tableScan} is present) or intermediate nodes
+ * ({@link CanonicalSubplan#childSubplan} is present).
+ * Canonical subplan nodes map to a plan subgraph for which a common subplan can be extracted.
+ * For instance, a canonical subplan can be extracted for "scan -> filter -> project" chain of operators
+ * or for "aggregation" operator.
+ * Canonical symbol names are derived from {@link CacheColumnId} (if symbol represents
+ * {@link ColumnHandle}). {@link CacheColumnId} for complex projections will use canonicalized and formatted
+ * version of projection expression.
+ */
+public class CanonicalSubplan
```

Comment:
Why do we need a dedicated class for this? Why isn't a regular PlanNode tree sufficient to represent a canonical structure (after canonicalization, of course)?

Reply:
This object contains more information than just the plan. For example, it contains `keyChain` that is later on used to determine if 2 (or more) `canonicalSubplan`s might be represented by the same subQuery (+ adaptation for each)

Reply:
> PlanNode itself will never be fully canonical (even after canonicalisation of symbols).

Why not? Can you give me an example? At the end of the day, canonicalization is about establishing a set of conventions.

Reply:
The examples are in the comment above. Essentially one needs to canonicalize group of nodes rather than a single `PlanNode` itself:

* `Scan` `Filter` `Project` are order invariant
* `Scan`, `ScanProject`, `ScanFilter`, `ScanFilterProject`, ... can be canonicalized to single `CanonicalSubplan` node, so that we can find common subplan and adaptations between subplans.
* Same as above, but for `Filter`, `FilterProject`, `ProjectFilter`, ...
* multi-joins can be canonicalized into single multi-join `CanonicalSubplan` node since join order is irrelevant.